src/tests/Microsoft.Diagnostics.NETCore.Client/KeyedHandleableCollectionTests.cs

```diff
@@ -0,0 +1,443 @@
+
+            // Check that the handler task didn't complete
+            Assert.Equal(delayTask, completedTask);
+
+            // Handler should have removed (2, "two") from the collection
+            Assert.Equal(4, collection.Count());
+            Assert.Equal(new string[] { "zero", "one", "three", "four" }, collection.Values);
+
+            // Handler should complete on this pair
+            const string expectedValue = "one + one";
+            collection.Add(2, expectedValue);
+
+            delayTask = Task.Delay(DefaultHandleTimeout);
+            completedTask = await Task.WhenAny(delayTask, handledValueTask);
+
+            // Check that the handler did complete
+            Assert.Equal(handledValueTask, completedTask);
+
+            // Check that it got the correct value
+            string handledValue = await handledValueTask;
+            Assert.Equal(expectedValue, handledValue);
+
+            // Check that handler did not remove the value
+            Assert.Equal(5, collection.Count());
+            Assert.Equal(new string[] { "zero", "one", "three", "four", expectedValue }, collection.Values);
+        }
+
+        [Fact]
+        public async Task KeyedHandleableCollectionHandleBeforeAddTest()
+        {
+            await KeyedHandleableCollectionHandleBeforeAddTestCore(useAsync: false);
+        }
+
+        [Fact]
+        public async Task KeyedHandleableCollectionHandleBeforeAddTestAsync()
+        {
+            await KeyedHandleableCollectionHandleBeforeAddTestCore(useAsync: true);
+        }
+
+        /// <summary>
+        /// Tests that handler can be added before an item is provided to the collection.
+        /// </summary>
+        private async Task KeyedHandleableCollectionHandleBeforeAddTestCore(bool useAsync)
+        {
+            using var collection = new TestKeyedHandleableCollection<int, string>();
+            Assert.Empty(collection);
+
+            var shim = new KeyedHandleableCollectionApiShim<int, string>(collection, useAsync);
+
+            // Register to be notified when handler is beginning to be processed
+            Task handlerBeginTask = collection.WaitForHandlerBeginAsync(DefaultHandleTimeout);
+
+            const string expectedValue = "three";
+
+            // Create task that will start handling BEFORE an item is added
+            Task<string> handleValueTask = Task.Run(() => shim.Handle(3, DefaultHandleTimeout));
+
+            // Wait for handler to begin processing
+            await handlerBeginTask;
+
+            IList<(int, string, int)> itemsAndCounts = new List<(int, string, int)>()
+            {
+                (0, "zero", 1),
+                (1, "one", 2),
+                (2, "two", 3),
+                (3, "three", 3), // Item is consumed immediately, thus collection count does not change
+                (4, "four", 4),
+                (5, "five", 5)
+            };
+            AddAndVerifyItems(collection, itemsAndCounts);
```

Comment:
This test is failing because the `handleValueTask` has probably timed out before reaching here and so it never consumes the 3rd item in the collection and causes the verify to fail.

Reply:
Increasing the timeout seems to have fixed this.